mathphys/functions.py

```diff
@@ -197,3 +286,60 @@ def get_package_string(package):
     else:
         repo_str += f'{ver:s}'
     return repo_str
+
+
+# ------------------------- HELPER METHODS ------------------------------------
+_TYPES2SAVE = (int, float, complex, str, bytes, bool)
+_STRTYPES = {typ.__name__ for typ in _TYPES2SAVE}
+_NPTYPES = (_np.int_, _np.float_, _np.complex_, _np.bool_)
+
+
+def _save_recursive_hdf5(fil, path, obj, compress):
+    typ = type(obj)
+    if isinstance(obj, _np.ndarray):
+        fil.create_dataset(path, data=obj, compression=compress)
+    elif isinstance(obj, str):  # Handle SiriusPVName
+        fil[path] = str(obj)
+    elif isinstance(obj, _TYPES2SAVE + _NPTYPES):
+        fil[path] = obj
+    elif obj is None:
+        fil[path] = 'None'
+    elif typ in {list, tuple, set}:
+        path = path.strip('/') + '/'
+        for i, el in enumerate(obj):
+            _save_recursive_hdf5(fil, path + f'{i:d}', el, compress)
+    elif typ == dict:
+        path = path.strip('/') + '/'
+        for key, item in obj.items():
+            _save_recursive_hdf5(fil, path + key, item, compress)
+    else:
+        raise TypeError('Data type not valid: '+typ.__name__+'.')
```

Comment:
Would it make sense to have this as a warning instead? Or option to make it a warning...

That way, if the data with unsupported data type isn't as important, it can simply be skipped.

Reply:
But, if the data isn't important, the user can remove it and call the method again. I think I prefer to let the function interface leaner.

Reply:
I'm thinking such a thing could require poking at internals of some apsuite class in order to change the data before it's saved.

However, given that you have already converted a lot of kinds of data into hdf5, it seems hitting such cases would be unlikely, so I'm okay with keeping the interface leaner.

Reply:
Humm, if that's what it takes, than that's what will have to be done :laughing: 
I think the ideia here is to try to establish a standard way of saving files. If there is some structure that does not follow this standard we either have to extend it or change that structure. In my next commits I'll even remove the support for saving and loading `SiriusPVName` objects, they will be converted to `str`. Talking to @xresende, it became clear that it is not a good idea to have that conditional import of `siriuspy` here. So in the future we will need to change the `apsuite` classes not to save `SiriusPVName`.